### Olympics Data Analysis ⚽

##### Download  Datset Using Kaggle ApI


In [1]:
# importing important libraries

import pandas as pd
import numpy as np
import os
import opendatasets as od
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px # data visualization

In [2]:
# Assign the Kaggle data set URL into variable
dataset = 'https://www.kaggle.com/datasets/arjunprasadsarkhel/2021-olympics-in-tokyo/data'
# Using opendatasets let's download the data sets
od.download(dataset)

Skipping, found downloaded files in ".\2021-olympics-in-tokyo" (use force=True to force download)


#### Lets talk about Describtion of the dataset after that we will download the dataset using Kaggle API

The Olympics is a major international sporting event that takes place every four years, bringing together athletes from around the world to compete in various sports. The modern Olympic Games were inspired by the ancient Olympic Games held in ancient Greece from the 8th century BCE to the 4th century CE.

The modern Olympics were first held in Athens, Greece, in 1896, and have since become the most prestigious and widely watched sporting event globally. The Games consist of two main editions: the Summer Olympics and the Winter Olympics, each held every four years but staggered, so there is an Olympic Games event every two years.

Key aspects of the Olympics include:

Participation: Athletes from different countries compete in the Games, fostering camaraderie, friendship, and understanding among nations.
Sports: A wide range of sports are featured in the Olympics, including athletics, swimming, gymnastics, soccer, basketball, skiing, ice hockey, and many others.

Host Cities: The Games are hosted by different cities around the world, selected through a competitive bidding process overseen by the International Olympic Committee (IOC).

Olympic Symbols: The Olympic Rings (five interlocking rings representing the continents) and the Olympic Flame (symbolizing the continuity between ancient and modern Games) are iconic symbols of the Olympics.

Olympic Values: The Games are guided by core principles, including excellence, friendship, respect, and fair play.

Ceremonies: The opening and closing ceremonies are grand spectacles that showcase the host country's culture and history.

Legacy: Hosting the Olympics often leaves a lasting legacy of improved infrastructure, sports facilities, and urban development in the host city.

The Olympics have faced challenges and controversies throughout their history, such as boycotts, doping scandals, and financial concerns. Nevertheless, they continue to captivate the world and inspire athletes to achieve their best on the global stage. The Olympics remain a celebration of human athleticism, international cooperation, and the pursuit of sporting excellence.

In the year 2020, Olympics had to happen in the Tokyo city of Japan but due to Covid-19 pandemic they got postponed to the year 2021. Here is an exploratory data analysis for the same olympics games held in the city of Tokyo in 2021.

#### loading the data

In [3]:

df_athletes = pd.read_excel("2021-olympics-in-tokyo\Athletes.xlsx")
df_coaches = pd.read_excel("2021-olympics-in-tokyo\Coaches.xlsx")
df_entries_gender = pd.read_excel("2021-olympics-in-tokyo\EntriesGender.xlsx")
df_medals = pd.read_excel("2021-olympics-in-tokyo\Medals.xlsx")
df_teams = pd.read_excel("2021-olympics-in-tokyo\Teams.xlsx")


g:\Summer Course\Advanced Ai\myenv\Lib\site-packages\openpyxl\styles\stylesheet.py:237: UserWarning: Workbook contains no default style, apply openpyxl's default
  warn("Workbook contains no default style, apply openpyxl's default")
g:\Summer Course\Advanced Ai\myenv\Lib\site-packages\openpyxl\styles\stylesheet.py:237: UserWarning: Workbook contains no default style, apply openpyxl's default
  warn("Workbook contains no default style, apply openpyxl's default")


#### Checking info for each part of Datset  Athletes Data

In [4]:
df_athletes.head(5)  # first 5 rows of our data


Name     NOC           Discipline
0    AALERUD Katrine  Norway         Cycling Road
1        ABAD Nestor   Spain  Artistic Gymnastics
2  ABAGNALE Giovanni   Italy               Rowing
3     ABALDE Alberto   Spain           Basketball
4      ABALDE Tamara   Spain           Basketball

In [5]:

df_athletes.info() # information about the data



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11085 entries, 0 to 11084
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Name        11085 non-null  object
 1   NOC         11085 non-null  object
 2   Discipline  11085 non-null  object
dtypes: object(3)
memory usage: 259.9+ KB


In [6]:
df_athletes.nunique() # number of unique values in each column

Name          11062
NOC             206
Discipline       46
dtype: int64

In [7]:

df_athletes.duplicated().sum() # number of duplicated rows



np.int64(1)

In [8]:
df_athletes.isnull().sum() # number of missing values in each column

Name          0
NOC           0
Discipline    0
dtype: int64

- There are three columns in the athletes dataset i.e., Name, NOC, Discipline.

- Name has all unique entries of name of all the participants taking part in the olympics.

- NOC and Discipline are having categorical values of number of country participants come from and Discipline they are participating in, has more than one entries for different participant.

- There is one duplicate row and zero null row.

#### Checking info for each part of Datset  Gender Data

In [9]:
df_entries_gender.head()


Discipline  Female  Male  Total
0       3x3 Basketball      32    32     64
1              Archery      64    64    128
2  Artistic Gymnastics      98    98    196
3    Artistic Swimming     105     0    105
4            Athletics     969  1072   2041

In [10]:
print("Number of rows and columns are :", df_entries_gender.shape)

Number of rows and columns are : (46, 4)


In [11]:
print(df_entries_gender.info()) # information about the data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46 entries, 0 to 45
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Discipline  46 non-null     object
 1   Female      46 non-null     int64 
 2   Male        46 non-null     int64 
 3   Total       46 non-null     int64 
dtypes: int64(3), object(1)
memory usage: 1.6+ KB
None


In [12]:
df_entries_gender.nunique() # number of unique values in each column

Discipline    46
Female        38
Male          41
Total         41
dtype: int64

In [13]:
print(df_entries_gender.duplicated().sum()) # number of duplicated rows

0


In [14]:
print(df_entries_gender.isnull().sum()) # number of missing values in each column

Discipline    0
Female        0
Male          0
Total         0
dtype: int64


Above data has four columns where one has str entries whereas other three has integral entries.

- Data has zero duplicate and zero null row.

- Data has 46 rows and 4 columns.

- The command nunique gives us the unique value column has.

#### Checking Info for each part of Datset Medals Data

In [15]:
df_medals.head()


Rank                    Team/NOC  Gold  Silver  Bronze  Total  \
0     1    United States of America    39      41      33    113   
1     2  People's Republic of China    38      32      18     88   
2     3                       Japan    27      14      17     58   
3     4               Great Britain    22      21      22     65   
4     5                         ROC    20      28      23     71   

   Rank by Total  
0              1  
1              2  
2              5  
3              4  
4              3

In [16]:
print("Number of rows and columns are :", df_medals.shape)

Number of rows and columns are : (93, 7)


In [17]:
df_medals.info() # information about the data


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 93 entries, 0 to 92
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Rank           93 non-null     int64 
 1   Team/NOC       93 non-null     object
 2   Gold           93 non-null     int64 
 3   Silver         93 non-null     int64 
 4   Bronze         93 non-null     int64 
 5   Total          93 non-null     int64 
 6   Rank by Total  93 non-null     int64 
dtypes: int64(6), object(1)
memory usage: 5.2+ KB


In [18]:
df_medals.nunique() # number of unique values in each column

Rank             67
Team/NOC         93
Gold             14
Silver           17
Bronze           21
Total            30
Rank by Total    30
dtype: int64

In [19]:
df_medals.duplicated().sum() # number of duplicated rows


np.int64(0)

In [20]:
df_medals.isnull().sum() # number of missing values in each column

Rank             0
Team/NOC         0
Gold             0
Silver           0
Bronze           0
Total            0
Rank by Total    0
dtype: int64

- Above data gives the rank-wise medal record of each country.

- Note that USA is on the top followed by China and Japan.

- Data has zero duplicate and zero null row entry.

- Data has 93 rows and 7 columns

#### Checking info of coaches Data

In [21]:
df_coaches.head() # first 5 rows of our data

Name            NOC  Discipline Event
0  ABDELMAGID Wael          Egypt    Football   NaN
1        ABE Junya          Japan  Volleyball   NaN
2    ABE Katsuhiko          Japan  Basketball   NaN
3     ADAMA Cherif  Côte d'Ivoire    Football   NaN
4       AGEBA Yuya          Japan  Volleyball   NaN

In [22]:
print("Number of rows and columns are :", df_coaches.shape) # number of rows and columns

Number of rows and columns are : (394, 4)


In [23]:

df_coaches.info() # information about the data



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 394 entries, 0 to 393
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Name        394 non-null    object
 1   NOC         394 non-null    object
 2   Discipline  394 non-null    object
 3   Event       249 non-null    object
dtypes: object(4)
memory usage: 12.4+ KB


In [24]:
df_coaches.nunique() # number of unique values in each column


Name          381
NOC            61
Discipline      9
Event           6
dtype: int64

In [25]:

df_coaches.duplicated().sum() # number of duplicated rows



np.int64(1)

In [26]:
df_coaches.isnull().sum() # number of missing values in each column

Name            0
NOC             0
Discipline      0
Event         145
dtype: int64

- Above data has 394 rows and 4 columns.

- It has record of coaches from different countries and discipline.

- Data has one duplicate row and zero null row.

- All the columns have str entries.

#### Checking info  of Teams Data

In [27]:
print(df_teams.head())
print()
print("Number of rows and columns are :", df_teams.shape)
print()
print(df_teams.info())
print()
print(df_teams.nunique())
print()
print(df_teams.duplicated().sum())
print()
print(df_teams.isnull().sum())

      Name      Discipline                         NOC  Event
0  Belgium  3x3 Basketball                     Belgium    Men
1    China  3x3 Basketball  People's Republic of China    Men
2    China  3x3 Basketball  People's Republic of China  Women
3   France  3x3 Basketball                      France  Women
4    Italy  3x3 Basketball                       Italy  Women

Number of rows and columns are : (743, 4)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 743 entries, 0 to 742
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Name        743 non-null    object
 1   Discipline  743 non-null    object
 2   NOC         743 non-null    object
 3   Event       743 non-null    object
dtypes: object(4)
memory usage: 23.3+ KB
None

Name          146
Discipline     20
NOC            84
Event          36
dtype: int64

0

Name          0
Discipline    0
NOC           0
Event         0
dtype: int64


### Making Data Visualization for each part of Dataset 

#### Visualizing the count of participants coming from different countries.

In [28]:
x = df_athletes.NOC.value_counts()
px.bar(x[:20], title="Top 20 countries in terms of number of participants :")

ValueError: Mime type rendering requires nbformat>=4.2.0 but it is not installed

Figure({
    'data': [{'alignmentgroup': 'True',
              'hovertemplate': 'variable=count<br>NOC=%{x}<br>value=%{y}<extra></extra>',
              'legendgroup': 'count',
              'marker': {'color': '#636efa', 'pattern': {'shape': ''}},
              'name': 'count',
              'offsetgroup': 'count',
              'orientation': 'v',
              'showlegend': True,
              'textposition': 'auto',
              'type': 'bar',
              'x': array(['United States of America', 'Japan', 'Australia',
                          "People's Republic of China", 'Germany', 'France', 'Canada',
                          'Great Britain', 'Italy', 'Spain', 'ROC', 'Brazil', 'Netherlands',
                          'Republic of Korea', 'New Zealand', 'Poland', 'Argentina',
                          'South Africa', 'Mexico', 'Hungary'], dtype=object),
              'xaxis': 'x',
              'y': array([615, 586, 470, 401, 400, 377, 368, 366, 356, 324, 318, 291, 274, 223,
                          202, 195, 180, 171, 155, 155]),
              'yaxis': 'y'}],
    'layout': {'barmode': 'relative',
               'legend': {'title': {'text': 'variable'}, 'tracegroupgap': 0},
               'template': '...',
               'title': {'text': 'Top 20 countries in terms of number of participants :'},
               'xaxis': {'anchor': 'y', 'domain': [0.0, 1.0], 'title': {'text': 'NOC'}},
               'yaxis': {'anchor': 'x', 'domain': [0.0, 1.0], 'title': {'text': 'value'}}}
})

- Above plot shows top count of participants coming from different countries.

- USA send maximum number of participants followed by Japan and Australia.

#### Visualizing the count of coaches coming from different countries.

In [29]:
x = df_coaches.NOC.value_counts()
px.bar(x[:20], title="Top 20 countries in terms of number of coaches :")


ValueError: Mime type rendering requires nbformat>=4.2.0 but it is not installed

Figure({
    'data': [{'alignmentgroup': 'True',
              'hovertemplate': 'variable=count<br>NOC=%{x}<br>value=%{y}<extra></extra>',
              'legendgroup': 'count',
              'marker': {'color': '#636efa', 'pattern': {'shape': ''}},
              'name': 'count',
              'offsetgroup': 'count',
              'orientation': 'v',
              'showlegend': True,
              'textposition': 'auto',
              'type': 'bar',
              'x': array(['Japan', 'United States of America', 'Spain', 'Australia', 'Canada',
                          'Italy', 'Egypt', 'South Africa', 'ROC', "People's Republic of China",
                          'Argentina', 'Netherlands', 'Venezuela', 'France', 'Nigeria', 'Germany',
                          'New Zealand', 'Mexico', 'Great Britain', 'Republic of Korea'],
                         dtype=object),
              'xaxis': 'x',
              'y': array([35, 28, 28, 22, 16, 14, 12, 12, 12, 12, 11, 10, 10, 10,  9,  9,  8,  8,
                           7,  7]),
              'yaxis': 'y'}],
    'layout': {'barmode': 'relative',
               'legend': {'title': {'text': 'variable'}, 'tracegroupgap': 0},
               'template': '...',
               'title': {'text': 'Top 20 countries in terms of number of coaches :'},
               'xaxis': {'anchor': 'y', 'domain': [0.0, 1.0], 'title': {'text': 'NOC'}},
               'yaxis': {'anchor': 'x', 'domain': [0.0, 1.0], 'title': {'text': 'value'}}}
})

- Above plot shows countries from where maximum number of coaches are coming from.

- Japan contributes to maximum number of coaches followed by USA and Spain.

#### Top Countries with most number of medals

In [30]:
px.bar(df_medals, x="Team/NOC", y="Total", color="Gold", title="Top Countries in terms of number of medals : ")

ValueError: Mime type rendering requires nbformat>=4.2.0 but it is not installed

Figure({
    'data': [{'alignmentgroup': 'True',
              'hovertemplate': 'Team/NOC=%{x}<br>Total=%{y}<br>Gold=%{marker.color}<extra></extra>',
              'legendgroup': '',
              'marker': {'color': array([39, 38, 27, 22, 20, 17, 10, 10, 10, 10,  7,  7,  7,  7,  6,  6,  4,  4,
                                          4,  4,  4,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  2,  2,  2,  2,
                                          2,  2,  2,  2,  2,  2,  2,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,
                                          1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  0,  0,  0,  0,  0,  0,  0,
                                          0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
                                          0,  0,  0]),
                         'coloraxis': 'coloraxis',
                         'pattern': {'shape': ''}},
              'name': '',
              'offsetgroup': '',
              'orientation': 'v',
              'showlegend': False,
              'textposition': 'auto',
              'type': 'bar',
              'x': array(['United States of America', "People's Republic of China", 'Japan',
                          'Great Britain', 'ROC', 'Australia', 'Netherlands', 'France', 'Germany',
                          'Italy', 'Canada', 'Brazil', 'New Zealand', 'Cuba', 'Hungary',
                          'Republic of Korea', 'Poland', 'Czech Republic', 'Kenya', 'Norway',
                          'Jamaica', 'Spain', 'Sweden', 'Switzerland', 'Denmark', 'Croatia',
                          'Islamic Republic of Iran', 'Serbia', 'Belgium', 'Bulgaria', 'Slovenia',
                          'Uzbekistan', 'Georgia', 'Chinese Taipei', 'Turkey', 'Greece', 'Uganda',
                          'Ecuador', 'Ireland', 'Israel', 'Qatar', 'Bahamas', 'Kosovo', 'Ukraine',
                          'Belarus', 'Romania', 'Venezuela', 'India', 'Hong Kong, China',
                          'Philippines', 'Slovakia', 'South Africa', 'Austria', 'Egypt',
                          'Indonesia', 'Ethiopia', 'Portugal', 'Tunisia', 'Estonia', 'Fiji',
                          'Latvia', 'Thailand', 'Bermuda', 'Morocco', 'Puerto Rico', 'Colombia',
                          'Azerbaijan', 'Dominican Republic', 'Armenia', 'Kyrgyzstan', 'Mongolia',
                          'Argentina', 'San Marino', 'Jordan', 'Malaysia', 'Nigeria', 'Bahrain',
                          'Saudi Arabia', 'Lithuania', 'North Macedonia', 'Namibia',
                          'Turkmenistan', 'Kazakhstan', 'Mexico', 'Finland', 'Botswana',
                          'Burkina Faso', "Côte d'Ivoire", 'Ghana', 'Grenada', 'Kuwait',
                          'Republic of Moldova', 'Syrian Arab Republic'], dtype=object),
              'xaxis': 'x',
              'y': array([113,  88,  58,  65,  71,  46,  36,  33,  37,  40,  24,  21,  20,  15,
                           20,  20,  14,  11,  10,   8,   9,  17,   9,  13,  11,   8,   7,   9,
                            7,   6,   5,   5,   8,  12,  13,   4,   4,   3,   4,   4,   3,   2,
                            2,  19,   7,   4,   4,   7,   6,   4,   4,   3,   7,   6,   5,   4,
                            4,   2,   2,   2,   2,   2,   1,   1,   1,   5,   7,   5,   4,   3,
                            4,   3,   3,   2,   2,   2,   1,   1,   1,   1,   1,   1,   8,   4,
                            2,   1,   1,   1,   1,   1,   1,   1,   1]),
              'yaxis': 'y'}],
    'layout': {'barmode': 'relative',
               'coloraxis': {'colorbar': {'title': {'text': 'Gold'}},
                             'colorscale': [[0.0, '#0d0887'], [0.1111111111111111,
                                            '#46039f'], [0.2222222222222222,
                                            '#7201a8'], [0.3333333333333333,
                                            '#9c179e'], [0.4444444444444444,
                                            '#bd3786'], [0.5555555555555556,
           

- Above plot shows highest ranked countries as per their medal count.

- USA is on the top follwed by China and Japan.

Note that Germany and Australia are in top 5 countries participants are coming from but not in top 5 countries having maximum medals.

- Whereas USA, China, Japan holds top position in both number of participants and medals won.

In [34]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, confusion_matrix

# Data Preparation for Logistic Regression
df_medals['Medal_Won'] = df_medals[['Gold', 'Silver', 'Bronze']].sum(axis=1) > 0
df_medals['Medal_Won'] = df_medals['Medal_Won'].astype(int)

# Merge the datasets
data = pd.merge(df_teams, df_medals[['Team/NOC', 'Medal_Won']], left_on='NOC', right_on='Team/NOC', how='left')
data['Medal_Won'].fillna(0, inplace=True)

# Encode categorical variables
label_encoder = LabelEncoder()
data['Discipline'] = label_encoder.fit_transform(data['Discipline'])
data['Event'] = label_encoder.fit_transform(data['Event'])

# Select features and target
X = data[['Discipline', 'Event']]
y = data['Medal_Won']

# Split the data into training and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Train Logistic Regression model with class balancing
logreg = LogisticRegression(class_weight='balanced')
logreg.fit(X_train, y_train)

# Predictions and evaluation
y_pred_logreg = logreg.predict(X_test)
print("Logistic Regression Classification Report:\n", classification_report(y_test, y_pred_logreg))
print("Logistic Regression Confusion Matrix:\n", confusion_matrix(y_test, y_pred_logreg))


Logistic Regression Classification Report:
               precision    recall  f1-score   support

         0.0       0.01      0.20      0.02         5
         1.0       0.97      0.56      0.71       218

    accuracy                           0.55       223
   macro avg       0.49      0.38      0.36       223
weighted avg       0.95      0.55      0.69       223

Logistic Regression Confusion Matrix:
 [[  1   4]
 [ 97 121]]


C:\Users\Mega\AppData\Local\Temp\ipykernel_16052\4001669545.py:12: FutureWarning:

A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.





#### Random Forest Classifier


In [35]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import classification_report, confusion_matrix

# Data Preparation for Random Forest
df_medals['Medal_Type'] = df_medals[['Gold', 'Silver', 'Bronze']].idxmax(axis=1)

# Merge the datasets
rf_data = pd.merge(df_teams, df_medals[['Team/NOC', 'Medal_Type']], left_on='NOC', right_on='Team/NOC', how='left')
rf_data.dropna(subset=['Medal_Type'], inplace=True)

# Encode categorical variables
rf_data['Discipline'] = label_encoder.fit_transform(rf_data['Discipline'])
rf_data['Event'] = label_encoder.fit_transform(rf_data['Event'])
rf_data['Medal_Type'] = label_encoder.fit_transform(rf_data['Medal_Type'])

# Select features and target
X_rf = rf_data[['Discipline', 'Event']]
y_rf = rf_data['Medal_Type']

# Split the data into training and test sets
X_rf_train, X_rf_test, y_rf_train, y_rf_test = train_test_split(X_rf, y_rf, test_size=0.3, random_state=42)

# Set up the parameter grid for hyperparameter tuning
param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

# Initialize Random Forest and perform grid search
rf_model = RandomForestClassifier(random_state=42)
grid_search = GridSearchCV(estimator=rf_model, param_grid=param_grid, cv=3, n_jobs=-1, verbose=2)
grid_search.fit(X_rf_train, y_rf_train)

# Best model from grid search
best_rf_model = grid_search.best_estimator_

# Predictions and evaluation
y_rf_pred = best_rf_model.predict(X_rf_test)
print("Random Forest Classification Report:\n", classification_report(y_rf_test, y_rf_pred, target_names=['Bronze', 'Gold', 'Silver']))
print("Random Forest Confusion Matrix:\n", confusion_matrix(y_rf_test, y_rf_pred))


Fitting 3 folds for each of 108 candidates, totalling 324 fits
Random Forest Classification Report:
               precision    recall  f1-score   support

      Bronze       0.42      0.66      0.51        97
        Gold       0.14      0.08      0.10        53
      Silver       0.24      0.13      0.17        69

    accuracy                           0.35       219
   macro avg       0.27      0.29      0.26       219
weighted avg       0.30      0.35      0.30       219

Random Forest Confusion Matrix:
 [[64 14 19]
 [40  4  9]
 [49 11  9]]


In [36]:
import pickle

# Save the best Random Forest model as a pickle file
with open('best_rf_model.pkl', 'wb') as file:
    pickle.dump(best_rf_model, file)


In [37]:
# Load the saved Random Forest model
with open('best_rf_model.pkl', 'rb') as file:
    loaded_rf_model = pickle.load(file)
# predictions
y_rf_pred_loaded = loaded_rf_model.predict(X_rf_test)
print("Loaded Random Forest Classification Report:\n", classification_report(y_rf_test, y_rf_pred_loaded, target_names=['Bronze', 'Gold', 'Silver']))

Loaded Random Forest Classification Report:
               precision    recall  f1-score   support

      Bronze       0.42      0.66      0.51        97
        Gold       0.14      0.08      0.10        53
      Silver       0.24      0.13      0.17        69

    accuracy                           0.35       219
   macro avg       0.27      0.29      0.26       219
weighted avg       0.30      0.35      0.30       219

